compiler/src/iree/compiler/Codegen/Common/ResolveWorkgroupCountHints.cpp

```diff
@@ -0,0 +1,1043 @@
+  }
+
+  // Subtract off any required operands that are also marked as optional
+  // operands.
+  slice.optionalOperands.set_subtract(slice.requiredOperands);
+
+  // Mark the function slice as valid. This way if the overall walk revisits it
+  // we will skip reprocessing it.
+  slice.valid = true;
+
+  // If this function transitively calls no hints then it isn't required to
+  // construct the workgroup count region. This culls processing of calls to
+  // this function.
+  if (slice.hints.empty() && slice.calls.empty()) {
+    slice.required = false;
+  }
+  return success();
+}
+
+//===---------------------------------------------------------------------===//
+// Workgroup Count Materialization
+//===---------------------------------------------------------------------===//
+
+/// Materializes a program slice. All |ordinals| are remapped to corresponding
+/// values in |workloadVals|. All dependent function arguments should already
+/// have been remapped and can be queried directly from |map|.
+static LogicalResult materializeSlice(
+    RewriterBase &rewriter, IRMapping &map, ValueRange workloadVals,
+    SetVector<Operation *> &slice,
+    SetVector<IREE::TensorExt::DispatchWorkloadOrdinalOp> &ordinals) {
+  for (auto ordinalOp : ordinals) {
+    // Map `tensor_ext.dispatch.workload.ordinal` op with the corresponding
+    // operand of the `tensor_ext.dispatch.workgroup_count_default` operation.
+    int64_t ordinal = ordinalOp.getOrdinal().getSExtValue();
+    if (ordinal >= workloadVals.size()) {
+      return ordinalOp.emitOpError(
+          "ordinal number is higher than the number of workloads captured in "
+          "the workgroup count region");
+    }
+    map.map(ordinalOp.getResult(), workloadVals[ordinal]);
+  }
+  for (auto op : slice) {
+    // TODO(#13038) This is a WAR for ops ending up in workgroup count
+    // computation. They should not. Some pre-processing at MaterializeEncoding
+    // time might make these go away.
+    if (isa<IREE::Codegen::QueryTileSizesOp>(op)) {
+      Value constVal =
+          arith::ConstantIndexOp::create(rewriter, op->getLoc(), 16);
+      for (auto result : op->getResults()) {
+        map.map(result, constVal);
+      }
+      continue;
+    }
+    rewriter.clone(*op, map);
+  }
+  return success();
+}
+
+/// Materializes a single combined `scf.if` based on the list of
+/// ConditionSlices. If no slices successfully materialize, this produces no
+/// new IR.
+static FailureOr<Operation *>
+materializeConditions(RewriterBase &rewriter, IRMapping &map, Location loc,
+                      ValueRange workloadVals,
+                      std::vector<ConditionSlice> &conditions) {
+  Value acc;
+  for (ConditionSlice &condition : conditions) {
+    if (!llvm::all_of(condition.funcArgLeaves,
+                      [&](BlockArgument b) { return map.contains(b); })) {
+      continue;
+    }
+
+    if (failed(materializeSlice(rewriter, map, workloadVals, condition.slice,
```

Comment:
Why reimplement this? There is already a method https://github.com/iree-org/iree/blob/2538af386c9400187899127584d475de65c16cf8/compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp#L425 that is supposed to do this. Why not use that?